lib/rules/no-unreachable.js

```diff
@@ -183,10 +189,29 @@ module.exports = {
             // Manages the current code path.
             onCodePathStart(codePath) {
                 currentCodePath = codePath;
+                codePathSegments.push(currentCodePathSegments);
+                currentCodePathSegments = [];
             },
 
             onCodePathEnd() {
                 currentCodePath = currentCodePath.upper;
+                currentCodePathSegments = codePathSegments.pop();
+            },
+
+            onUnreachableCodePathSegmentStart(segment) {
+                currentCodePathSegments.push(segment);
+            },
+
+            onUnreachableCodePathSegmentEnd() {
+                currentCodePathSegments.pop();
+            },
+
+            onCodePathSegmentStart(segment) {
+                currentCodePathSegments.push(segment);
+            },
+
+            onCodePathSegmentEnd() {
+                currentCodePathSegments.pop();
             },
```

Comment:
Is it guaranteed that the ending code path segment is always the last in this list?

Reply:
I believe so. `onCodePathSegmentEnd` and `onUnreachableCodePathSegmentEnd` fire in the opposite order of `onCodePathSegmentStart` and `onUnreachableCodePathSegmentStart`, so the most recent segment pushed onto the stack should be the first popped off.

Reply:
For this code:

```js
try { foo(); }
finally
{
    try { bar(); }
    finally { baz(); }
}
```

I'm getting events fired in this order:

<pre>
s1_1: onCodePathSegmentStart
s1_1: onCodePathSegmentEnd
s1_2: onCodePathSegmentStart
s1_2: onCodePathSegmentEnd
s1_3: onCodePathSegmentStart
s1_4: onCodePathSegmentStart
s1_3: onCodePathSegmentEnd
s1_4: onCodePathSegmentEnd
s1_5: onCodePathSegmentStart
s1_6: onCodePathSegmentStart
s1_5: onCodePathSegmentEnd
s1_6: onCodePathSegmentEnd
s1_7: onCodePathSegmentStart
s1_8: onCodePathSegmentStart
s1_9: onCodePathSegmentStart
s1_10: onCodePathSegmentStart
s1_9: onCodePathSegmentEnd
s1_10: onCodePathSegmentEnd
s1_8: onCodePathSegmentEnd
s1_7: onCodePathSegmentEnd
</pre>

where segments `s1_9` and `s1_10` are firing `onCodePathSegmentEnd` in reverse order. I'm not sure though if this behavior is intentional or if changing it would be a problem.

Reply:
That's interesting! This code is the culprit:

https://github.com/eslint/eslint/blob/926a28684282aeec37680bbc52a66973b8055f54/lib/linter/code-path-analysis/code-path-analyzer.js#L189-L205

My best guess is that this is a bug. I tried switching the order of events and none of our tests fail. 

I suppose the question is if we should fix it in this PR or not? For the way we are using these events, the order in which they fire isn't relevant, and my guess is that this will be true for anyone currently using these events.

Reply:
It's difficult to say whether this is a bug. When multiple segments are "current" at the same time, they're not nested within each other (in which case it would definitely make the most sense for them to end in reverse order), but rather alternative to each other, so the current behavior where alternative paths are always processed in the same order might make more sense than reversing the order in which alternative segments end.

Reply:
Here's the graph for https://github.com/eslint/eslint/pull/17511#discussion_r1311179774, I just added s1_1, s1_2... for clarity:

[https://dreampuf.github.io/GraphvizOnline/#digraph...](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0D%0Anode%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0D%0Ainitial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0D%0Afinal%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0D%0Athrown%5Blabel%3D%22%E2%9C%98%22%2Cshape%3Dcircle%2Cwidth%3D0.3%2Cheight%3D0.3%2Cfixedsize%3Dtrue%5D%3B%0D%0As1_1%5Blabel%3D%22s1_1%5CnProgram%3Aenter%5CnTryStatement%3Aenter%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(foo)%22%5D%3B%0D%0As1_2%5Blabel%3D%22s1_2%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0D%0As1_3%5Blabel%3D%22s1_3%5CnBlockStatement%3Aenter%5CnTryStatement%3Aenter%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(bar)%22%5D%3B%20%20%20%20%20%20%20%20%0D%0As1_5%5Blabel%3D%22s1_5%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0D%0As1_7%5Blabel%3D%22s1_7%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(baz)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%5CnTryStatement%3Aexit%5CnBlockStatement%3Aexit%5CnTryStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0D%0As1_4%5Blabel%3D%22s1_4%5CnBlockStatement%3Aenter%5CnTryStatement%3Aenter%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(bar)%22%5D%3B%20%20%20%20%20%20%20%20%0D%0As1_6%5Blabel%3D%22s1_6%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0D%0As1_8%5Blabel%3D%22s1_8%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(baz)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%5CnTryStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0D%0As1_9%5Blabel%3D%22s1_9%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(baz)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0D%0As1_10%5Blabel%3D%22s1_10%5CnBlockStatement%3Aenter%5CnExpressionStatement%3Aenter%5CnCallExpression%3Aenter%5CnIdentifier%20(baz)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0D%0Ainitial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_5-%3Es1_7%3B%0D%0As1_1-%3Es1_4-%3Es1_6-%3Es1_8%3B%0D%0As1_2-%3Es1_4%3B%0D%0As1_3-%3Es1_9%3B%0D%0As1_5-%3Es1_9%3B%0D%0As1_4-%3Es1_10%3B%0D%0As1_6-%3Es1_10%3B%0D%0As1_7-%3Efinal%3B%0D%0As1_9-%3Ethrown%3B%0D%0As1_10-%3Ethrown%3B%0D%0As1_8-%3Ethrown%3B%0D%0A%7D)

In nested `try` statements, alternative segments in `finally` blocks that relate to throwing/returning from its `try` all start after all segments that relate to not throwing/returning from its `try`. Since the former ones lead to exiting right after `finally`, they can only end before (when they're the last segments of `finally)` or at the same time as the latter ones, so it seems that segments that start later will indeed always end before or at the same time as segments that started earlier.

> I suppose the question is if we should fix it in this PR or not? For the way we are using these events, the order in which they fire isn't relevant, and my guess is that this will be true for anyone currently using these events.

I think we shouldn't change the order in which the events fire.

I also believe that `onCodePathSegmentEnd() { currentCodePathSegments.pop(); }` will work fine for our rules even in cases when it removes a wrong segment for that particular call because another `onCodePathSegmentEnd` will fire right after. However, for the sake of clarity, and in case we missed some cases where this wouldn't work well, it might be good to update this code to look for and remove the exact segment for which the event was fired.

Reply:
Sounds good. I'll make the change.